File: src/export/flatten.ts
import type { Attribution, CheckerResult, ToolExecutionMetrics, TrajectoryIR, Violation } from "../types/index.js";
import type { CursorEvent } from "../types/index.js";
import { formatDuration, formatTokenCount } from "../enrich/toolMetrics.js";
import { resolveSessionActiveWallMs, resolveSessionWallMs, resolveUserIdleMs } from "../ir/sessionMetrics.js";

const USER_QUERY_RE = /<user_query>\s*(.*?)\s*<\/user_query>/s;
const REDACTED_RE = /\[REDACTED\]/g;

function extractUserText(raw: string): string {
  const m = USER_QUERY_RE.exec(raw);
  let text = m ? m[1]! : raw;
  text = text.replace(/<image_files>[\s\S]*?<\/image_files>/g, "").replace(/\[Image\]\s*/g, "");
  return text.trim();
}

function truncate(s: string, max: number): string {
  if (max <= 0 || s.length <= max) return s;
  return s.slice(0, max - 20) + "\n\n… [truncated]";
}

function formatExecutionMetrics(exec?: ToolExecutionMetrics): string[] {
  if (!exec) return [];
  const lines: string[] = [];
  lines.push(`- **duration:** ${formatDuration(exec.duration_ms)} (${exec.duration_source})`);
  lines.push(`- **output:** ${formatTokenCount(exec.output_chars)} chars / ~${formatTokenCount(exec.output_tokens)} tokens (${exec.output_source})`);
  if (exec.output_path) lines.push(`- **output_path:** \`${exec.output_path}\``);
  return lines;
}

function formatToolBlock(name: string, inp: Record<string, unknown>, maxToolChars: number, exec?: ToolExecutionMetrics): string {
  const lines: string[] = [`### Tool: \`${name}\``, ""];
  lines.push(...formatExecutionMetrics(exec));
  if (lines.length > 2) lines.push("");

  if (name === "CallMcpTool") {
    lines.push(`- **MCP server:** \`${inp.server ?? "unknown"}\``);
    lines.push(`- **Tool:** \`${inp.toolName ?? "unknown"}\``);
    const args = inp.arguments as Record<string, unknown> | undefined;
    if (args && Object.keys(args).length) {
      lines.push("", "```json", truncate(JSON.stringify(args, null, 2), maxToolChars), "```");
    }
    return lines.join("\n");
  }

  if (name === "Shell") {
    if (inp.description) lines.push(`- ${inp.description}`);
    lines.push("", "```bash", truncate(String(inp.command ?? ""), maxToolChars), "```");
    return lines.join("\n");
  }

  if (name === "Read") {
    const extra: string[] = [];
    if (inp.offset != null) extra.push(`offset=${inp.offset}`);
    if (inp.limit != null) extra.push(`limit=${inp.limit}`);
    const suffix = extra.length ? ` (${extra.join(", ")})` : "";
    lines.push(`- **path:** \`${inp.path ?? ""}\`${suffix}`);
    return lines.join("\n");
  }

  if (name === "Grep") {
    lines.push(`- **pattern:** \`${inp.pattern ?? ""}\``);
    if (inp.path) lines.push(`- **path:** \`${inp.path}\``);
    if (inp.glob) lines.push(`- **glob:** \`${inp.glob}\``);
    return lines.join("\n");
  }

  if (name === "Write" || name === "StrReplace") {
    lines.push(`- **path:** \`${inp.path ?? ""}\``);
    return lines.join("\n");
  }

  if (name === "Glob") {
    lines.push(`- **pattern:** \`${inp.glob_pattern ?? ""}\``);
    if (inp.target_directory) lines.push(`- **dir:** \`${inp.target_directory}\``);
    return lines.join("\n");
  }

  lines.push("", "```json", truncate(JSON.stringify(inp, null, 2), maxToolChars), "```");
  return lines.join("\n");
}

export interface FlattenOptions {
  trajectoryId?: string;
  sourcePath?: string;
  maxToolChars?: number;
  maxAssistantChars?: number;
  toolMetrics?: Map<string, ToolExecutionMetrics>;
  sessionToolStats?: Record<string, unknown>;
}

export function flattenEventsToMarkdown(events: CursorEvent[], opts: FlattenOptions = {}): string {
  const {
    trajectoryId = "unknown",
    sourcePath = "",
    maxToolChars = 4000,
    maxAssistantChars = 8000,
    toolMetrics,
    sessionToolStats,
  } = opts;

  let userTurn = 0;
  let stepIdx = 0;
  let instruction = "";

  const parts: string[] = [
    "# Cursor Session Transcript",
    "",
    "## Metadata",
    "",
    "| Field | Value |",
    "|-------|-------|",
    `| session_id | \`${trajectoryId}\` |`,
  ];
  if (sourcePath) parts.push(`| source | \`${sourcePath}\` |`);
  parts.push(
    "",
    "> Flattened by **TrajRx** for AI attribution reading. Chronological, one section per user message or assistant step.",
    ""
  );

  const body: string[] = [];

  for (const event of events) {
    const role = event.role;
    const contentList = event.message?.content ?? [];

    if (role === "user") {
      userTurn++;
      const texts = contentList
        .filter((c) => c.type === "text")
        .map((c) => extractUserText(String(c.text ?? "")))
        .filter(Boolean);
      const userText = texts.join("\n\n").trim();
      if (userText && !instruction) instruction = userText.slice(0, 2000);
      body.push("---", "", `## User Turn ${userTurn} (#U${userTurn})`, "", userText || "_(empty user message)_", "");
      continue;
    }

    if (role !== "assistant") continue;

    stepIdx++;
    body.push("---", "", `## Assistant Step ${stepIdx} (#S${stepIdx}, after #U${userTurn})`, "");

    let subIdx = 0;
    for (const item of contentList) {
      if (item.type === "text") {
        const text = String(item.text ?? "").trim().replace(REDACTED_RE, "_[thinking redacted]_");
        if (text) body.push("", truncate(text, maxAssistantChars), "");
      } else if (item.type === "tool_use") {
        subIdx++;
        body.push("", formatToolBlock(
          String(item.name ?? "unknown"),
          (item.input ?? {}) as Record<string, unknown>,
          maxToolChars,
          toolMetrics?.get(`${stepIdx}:${subIdx}`)
        ), "");
      }
    }
  }

  if (instruction) {
    parts.push("## Task (first user message)", "", instruction, "");
  }
  parts.push("## Conversation", "", ...body);
  parts.push("---", "", "## Session Stats", "", `- user_turns: ${userTurn}`, `- assistant_steps: ${stepIdx}`, "");
  if (sessionToolStats) {
    parts.push("## Tool Efficiency Summary", "");
    parts.push(`- total_tool_duration: ${formatDuration(Number(sessionToolStats.total_duration_ms ?? 0))}`);
    parts.push(`- total_output_tokens: ~${formatTokenCount(Number(sessionToolStats.total_output_tokens ?? 0))}`);
    parts.push(`- total_output_chars: ${formatTokenCount(Number(sessionToolStats.total_output_chars ?? 0))}`);
    const slowest = (sessionToolStats.slowest as Array<{ step: number; sub_index: number; tool: string; duration_ms: number; output_tokens: number }>) ?? [];
    if (slowest.length) {
      parts.push("", "### Slowest tools", "");
      for (const row of slowest.slice(0, 5)) {
        parts.push(`- #S${row.step} \`${row.tool}\`: ${formatDuration(row.duration_ms)}, ~${formatTokenCount(row.output_tokens)} tokens`);
      }
    }
    const largest = (sessionToolStats.largest_outputs as Array<{ step: number; sub_index: number; tool: string; duration_ms: number; output_tokens: number }>) ?? [];
    if (largest.length) {
      parts.push("", "### Largest outputs", "");
      for (const row of largest.slice(0, 5)) {
        parts.push(`- #S${row.step} \`${row.tool}\`: ~${formatTokenCount(row.output_tokens)} tokens (${formatDuration(row.duration_ms)})`);
      }
    }
    parts.push("");
  }

  return parts.join("\n");
}

export function appendAttributionSection(flatMd: string, attribution: Attribution, violations?: Violation[]): string {
  const lines: string[] = [
    "",
    "---",
    "",
    "## Attribution Summary (TrajRx)",
    "",
    `- **primary_cause:** ${attribution.primary_cause}`,
    `- **confidence:** ${attribution.confidence}`,
    `- **critical_step:** ${attribution.critical_step}`,
    "",
  ];
  if (attribution.composite_causes?.length) {
    lines.push(`- **composite:** ${attribution.composite_causes.join(", ")}`, "");
  }
  const scores = attribution.category_scores ?? {};
  if (Object.keys(scores).length) {
    lines.push("### Category scores", "");
    for (const [cat, sc] of Object.entries(scores).sort((a, b) => b[1] - a[1])) {
      lines.push(`- ${cat}: ${sc}`);
    }
    lines.push("");
  }
  if (violations?.length) {
    lines.push("### Top violations", "");
    for (const v of violations.slice(0, 8)) {
      lines.push(`- \`[${v.severity}]\` **${v.invariant_id}** @ step ${v.step_index}: ${v.message}`);
    }
  }
  return flatMd + lines.join("\n");
}

export function buildReport(traj: TrajectoryIR, checker: CheckerResult, attr: Attribution): string {
  const tel = checker.telemetry_summary as Record<string, number | Record<string, number>>;
  const lines = [
    "# TrajRx Attribution Report",
    "",
    `**Session:** \`${traj.trajectory_id}\``,
    `**Source:** ${traj.source}`,
    "",
    "## Task",
    "",
    (traj.instruction || "(no instruction)").slice(0, 1500),
    "",
    "## Summary",
    "",
    "| Metric | Value |",
    "|--------|-------|",
    `| Primary cause | **${attr.primary_cause}** (confidence ${attr.confidence}) |`,
    `| Critical step | ${attr.critical_step} |`,
    `| Assistant steps | ${tel.step_count ?? 0} |`,
    `| User turns | ${tel.user_turns ?? 0} |`,
    resolveSessionWallMs(traj) != null ? `| Session wall time (gross) | ${formatDuration(resolveSessionWallMs(traj)!)} |` : null,
    resolveSessionActiveWallMs(traj) != null ? `| Session active wall time (net) | ${formatDuration(resolveSessionActiveWallMs(traj)!)} |` : null,
    resolveUserIdleMs(traj) != null && resolveUserIdleMs(traj)! > 0 ? `| User idle time | ${formatDuration(resolveUserIdleMs(traj)!)} |` : null,
    `| Violations | ${checker.violation_count} |`,
    "",
  ];
  if (tel.total_tool_duration_ms != null) {
    lines.push("## Tool Efficiency", "", "| Metric | Value |", "|--------|-------|");
    lines.push(`| Total tool time | ${formatDuration(Number(tel.total_tool_duration_ms))} |`);
    lines.push(`| Total output tokens | ~${formatTokenCount(Number(tel.total_output_tokens ?? 0))} |`);
    lines.push(`| Avg duration / tool | ${formatDuration(Number(tel.avg_tool_duration_ms ?? 0))} |`);
    lines.push(`| Avg output tokens / tool | ~${formatTokenCount(Math.round(Number(tel.avg_output_tokens_per_tool ?? 0)))} |`);
    lines.push("");
  }
  lines.push(
    "## Explanation",
    "",
    attr.explanation,
  );
  return lines.join("\n");
}

export function formatReconcileReport(rec: Record<string, unknown>): string {
  const lines = [
    "# Attribution Reconciliation",
    "",
    `**Verdict:** ${rec.verdict} (primary_match=${rec.primary_match})`,
    "",
    "| Method | Primary | Confidence |",
    "|--------|---------|------------|",
    `| Static | ${rec.static_primary} | ${rec.static_confidence} |`,
    `| Manual | ${rec.manual_primary} | ${rec.manual_confidence} |`,
    "",
  ];
  for (const n of (rec.notes as string[]) ?? []) lines.push(`- ${n}`);
  return lines.join("\n");
}

File: doctor/export/flatten.py
"""Flatten Cursor JSONL transcript to a single linear Markdown file for AI reading."""

from __future__ import annotations

import json
import os
import re
from typing import Any

USER_QUERY_RE = re.compile(r"<user_query>\s*(.*?)\s*</user_query>", re.DOTALL)
REDACTED_RE = re.compile(r"\[REDACTED\]")


def _extract_user_text(raw: str) -> str:
    m = USER_QUERY_RE.search(raw)
    if m:
        return m.group(1).strip()
    # strip image_files blocks for readability
    text = re.sub(r"<image_files>.*?</image_files>", "", raw, flags=re.DOTALL)
    text = re.sub(r"\[Image\]\s*", "", text)
    return text.strip()


def _truncate(s: str, max_chars: int) -> str:
    if max_chars <= 0 or len(s) <= max_chars:
        return s
    return s[: max_chars - 20] + "\n\n… [truncated]"


def _format_tool_block(name: str, inp: dict[str, Any], max_tool_chars: int) -> str:
    lines = [f"### Tool: `{name}`", ""]

    if name == "CallMcpTool":
        server = inp.get("server") or "unknown"
        tool = inp.get("toolName") or "unknown"
        args = inp.get("arguments") or {}
        lines.append(f"- **MCP server:** `{server}`")
        lines.append(f"- **Tool:** `{tool}`")
        if args:
            body = json.dumps(args, ensure_ascii=False, indent=2)
            lines.extend(["", "```json", _truncate(body, max_tool_chars), "```"])
        return "\n".join(lines)

    if name == "Shell":
        cmd = inp.get("command") or ""
        desc = inp.get("description") or ""
        if desc:
            lines.append(f"- {desc}")
        lines.extend(["", "```bash", _truncate(str(cmd), max_tool_chars), "```"])
        return "\n".join(lines)

    if name == "Read":
        path = inp.get("path") or ""
        offset = inp.get("offset")
        limit = inp.get("limit")
        extra = []
        if offset is not None:
            extra.append(f"offset={offset}")
        if limit is not None:
            extra.append(f"limit={limit}")
        suffix = f" ({', '.join(extra)})" if extra else ""
        lines.append(f"- **path:** `{path}`{suffix}")
        return "\n".join(lines)

    if name == "Grep":
        lines.append(f"- **pattern:** `{inp.get('pattern', '')}`")
        if inp.get("path"):
            lines.append(f"- **path:** `{inp.get('path')}`")
        if inp.get("glob"):
            lines.append(f"- **glob:** `{inp.get('glob')}`")
        return "\n".join(lines)

    if name in ("Write", "StrReplace"):
        lines.append(f"- **path:** `{inp.get('path', '')}`")
        if name == "StrReplace" and inp.get("old_string"):
            old = str(inp["old_string"])[:200]
            lines.append(f"- **old (preview):** `{old}…`" if len(str(inp["old_string"])) > 200 else f"- **old:** `{old}`")
        return "\n".join(lines)

    if name == "Glob":
        lines.append(f"- **pattern:** `{inp.get('glob_pattern', '')}`")
        if inp.get("target_directory"):
            lines.append(f"- **dir:** `{inp.get('target_directory')}`")
        return "\n".join(lines)

    if name == "Task":
        lines.append(f"- **subagent:** `{inp.get('subagent_type', '')}`")
        if inp.get("description"):
            lines.append(f"- **description:** {inp.get('description')}")
        return "\n".join(lines)

    body = json.dumps(inp, ensure_ascii=False, indent=2)
    lines.extend(["", "```json", _truncate(body, max_tool_chars), "```"])
    return "\n".join(lines)


def flatten_events_to_markdown(
    events: list[dict[str, Any]],
    *,
    trajectory_id: str = "unknown",
    source_path: str = "",
    max_tool_chars: int = 4000,
    max_assistant_chars: int = 8000,
    include_step_index: bool = True,
) -> str:
    """Convert raw Cursor JSONL events to flat chronological Markdown."""
    user_turn = 0
    step_idx = 0
    instruction = ""

    parts: list[str] = [
        f"# Cursor Session Transcript",
        "",
        "## Metadata",
        "",
        f"| Field | Value |",
        f"|-------|-------|",
        f"| session_id | `{trajectory_id}` |",
    ]
    if source_path:
        parts.append(f"| source | `{source_path}` |")
    parts.extend([
        "",
        "> Flattened by **doctor** for AI attribution reading. Chronological, one section per user message or assistant step.",
        "",
    ])

    body_parts: list[str] = []

    for event in events:
        role = event.get("role")
        msg = event.get("message") or {}
        content_list = msg.get("content") or []

        if role == "user":
            user_turn += 1
            texts = []
            for item in content_list:
                if item.get("type") == "text":
                    texts.append(_extract_user_text(str(item.get("text") or "")))
            user_text = "\n\n".join(t for t in texts if t).strip()
            if user_text and not instruction:
                instruction = user_text[:2000]

            body_parts.extend([
                "---",
                "",
                f"## User Turn {user_turn} (#U{user_turn})",
                "",
                user_text or "_(empty user message)_",
                "",
            ])
            continue

        if role != "assistant":
            continue

        step_idx += 1
        header = f"## Assistant Step {step_idx} (#S{step_idx}, user_turn={user_turn})"
        if include_step_index:
            header = f"## Assistant Step {step_idx} (#S{step_idx}, after #U{user_turn})"

        body_parts.extend(["---", "", header, ""])

        for item in content_list:
            t = item.get("type")
            if t == "text":
                text = str(item.get("text") or "").strip()
                text = REDACTED_RE.sub("_[thinking redacted]_", text)
                if text:
                    body_parts.extend(["", _truncate(text, max_assistant_chars), ""])
            elif t == "tool_use":
                name = str(item.get("name") or "unknown")
                inp = item.get("input") or {}
                body_parts.extend(["", _format_tool_block(name, inp, max_tool_chars), ""])

    if instruction:
        parts.extend([
            "## Task (first user message)",
            "",
            instruction,
            "",
        ])

    parts.extend(["## Conversation", ""])
    parts.extend(body_parts)

    # footer stats
    parts.extend([
        "---",
        "",
        "## Session Stats",
        "",
        f"- user_turns: {user_turn}",
        f"- assistant_steps: {step_idx}",
        "",
    ])

    return "\n".join(parts)


def flatten_file(
    input_path: str,
    output_path: str | None = None,
    *,
    max_tool_chars: int = 4000,
    max_assistant_chars: int = 8000,
) -> str:
    """Read JSONL file and write flat Markdown."""
    from doctor.ir.loader import load_trajectories

    raw = load_trajectories(input_path)
    traj = raw[0]
    events = traj.get("events") or []
    tid = traj.get("trajectory_id") or os.path.splitext(os.path.basename(input_path))[0]

    md = flatten_events_to_markdown(
        events,
        trajectory_id=tid,
        source_path=input_path,
        max_tool_chars=max_tool_chars,
        max_assistant_chars=max_assistant_chars,
    )

    if output_path is None:
        stem = os.path.splitext(os.path.basename(input_path))[0]
        output_path = os.path.join(os.path.dirname(input_path), f"{stem}.flat.md")

    os.makedirs(os.path.dirname(os.path.abspath(output_path)), exist_ok=True)
    with open(output_path, "w", encoding="utf-8") as f:
        f.write(md)
    return output_path


def append_attribution_section(md_path: str, attribution: dict[str, Any], violations: list[dict[str, Any]] | None = None) -> None:
    """Append static attribution summary to an existing flat markdown file."""
    lines = [
        "",
        "---",
        "",
        "## Attribution Summary (doctor)",
        "",
        f"- **primary_cause:** {attribution.get('primary_cause')}",
        f"- **confidence:** {attribution.get('confidence')}",
        f"- **critical_step:** {attribution.get('critical_step')}",
        "",
    ]
    if attribution.get("composite_causes"):
        lines.append(f"- **composite:** {', '.join(attribution['composite_causes'])}")
        lines.append("")

    scores = attribution.get("category_scores") or {}
    if scores:
        lines.append("### Category scores")
        lines.append("")
        for cat, sc in sorted(scores.items(), key=lambda x: -x[1]):
            lines.append(f"- {cat}: {sc}")
        lines.append("")

    if violations:
        lines.append("### Top violations")
        lines.append("")
        for v in violations[:8]:
            lines.append(f"- `[{v.get('severity')}]` **{v.get('invariant_id')}** @ step {v.get('step_index')}: {v.get('message')}")
        lines.append("")

    with open(md_path, "a", encoding="utf-8") as f:
        f.write("\n".join(lines))
